src/Workspaces/Core/Portable/Workspace/EventMap.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis;
+
+internal sealed class EventMap
+{
+    private readonly SemaphoreSlim _guard = new(initialCount: 1);
+    private readonly Dictionary<string, EventHandlerSet> _eventNameToHandlerSet = [];
+
+    public WorkspaceEventRegistration AddEventHandler(string eventName, WorkspaceEventHandlerAndOptions handlerAndOptions)
+    {
+        using (_guard.DisposableWait())
+        {
+            _eventNameToHandlerSet[eventName] = _eventNameToHandlerSet.TryGetValue(eventName, out var handlers)
+                ? handlers.AddHandler(handlerAndOptions)
+                : EventHandlerSet.Empty.AddHandler(handlerAndOptions);
+        }
+
+        return new WorkspaceEventRegistration(this, eventName, handlerAndOptions);
+    }
+
+    public void RemoveEventHandler(string eventName, WorkspaceEventHandlerAndOptions handlerAndOptions)
+    {
+        using (_guard.DisposableWait())
+        {
+            if (_eventNameToHandlerSet.TryGetValue(eventName, out var handlers))
+                _eventNameToHandlerSet[eventName] = handlers.RemoveHandler(handlerAndOptions);
+        }
+    }
+
+    public EventHandlerSet GetEventHandlerSet(string eventName)
+    {
+        using (_guard.DisposableWait())
+        {
+            return _eventNameToHandlerSet.TryGetValue(eventName, out var handlers)
+                ? handlers
+                : EventHandlerSet.Empty;
+        }
+    }
+
+    public record struct WorkspaceEventHandlerAndOptions(Action<EventArgs> Handler, WorkspaceEventOptions Options)
+    {
+    }
+
+    public sealed class EventHandlerSet(ImmutableArray<Registry> registries)
+    {
+        public static readonly EventHandlerSet Empty = new([]);
+        private readonly ImmutableArray<Registry> _registries = registries;
+
+        public static EventHandlerSet Create(WorkspaceEventHandlerAndOptions handlerAndOptions)
+            => new EventHandlerSet([new Registry(handlerAndOptions)]);
+
+        public EventHandlerSet AddHandler(WorkspaceEventHandlerAndOptions handlerAndOptions)
+            => new EventHandlerSet(_registries.Add(new Registry(handlerAndOptions)));
+
+        public EventHandlerSet RemoveHandler(WorkspaceEventHandlerAndOptions handlerAndOptions)
+        {
+            var newRegistries = _registries.RemoveAll(r => r.HasHandlerAndOptions(handlerAndOptions));
```

Comment:
i need docs on why this is RemoveAll, and not removing the first.  the lack  of symmetry here is worrying.  like you could add twice, and remove once, and be back to empty.

Reply:
I don't like that either (it's what the code did before), I think just removing the first matching item is better.